[GTK] imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https.html is crashing in a Skia codepath

nikolaszimmermann opened the pull request:
#### 02d6be0a697de0b2e7a1c5868b2e6eebe50cfb8a
<pre>
[GTK] imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https.html is crashing in a Skia codepath
<a href="https://bugs.webkit.org/show_bug.cgi?id=304134">https://bugs.webkit.org/show_bug.cgi?id=304134</a>

Reviewed by Adrian Perez de Castro.

When using Skia with GPU acceleration, NativeImages backed by textures must be
destroyed on the same thread where the Skia GrDirectContext was created, not on
the compositor thread. Releasing GPU resources on the wrong thread corrupts
Skia&apos;s GrResourceCache, leading to bounds check failures and crashes.

This patch ensures that in ~CoordinatedPlatformLayerBufferNativeImage(), if the
image is texture-backed, its destruction is dispatched to the main thread.

This also fixes crashes/assertions in:
- http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable.html
- http/wpt/mediastream/transfer-mediastreamtrack-to-worker.html
- imported/w3c/web-platform-tests/mst-content-hint/RTCRtpSendParameters-degradationEffect.html

Covered by existing tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedImageBackingStore.cpp:
(WebCore::CoordinatedImageBackingStore::isSameNativeImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.cpp:
(WebCore::CoordinatedPlatformLayerBufferNativeImage::~CoordinatedPlatformLayerBufferNativeImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.h:

Canonical link: <a href="https://commits.webkit.org/304897@main">https://commits.webkit.org/304897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00bb1a561c5ecd4d7314e580d94ccbb3495cad07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144378 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89626 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9713 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104497 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/75719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139598 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7096 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122445 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6740 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4459 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4973 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40633 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147137 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8696 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41207 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112851 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7317 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113180 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28785 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6662 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118737 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62812 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8744 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36791 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8464 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72310 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8684 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8536 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

csaavedra commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -2744,10 +2744,6 @@ imported/w3c/web-platform-tests/webrtc/protocol/codecs-subsequent-offer.https.ht
 
-# The 'Can demux two video tracks with the same payload type on an unbundled connection' test times
-# out, the metadataToBeLoaded promise never resolves. The skia crash seems to be a consequence of the timeout.
-webkit.org/b/286481 imported/w3c/web-platform-tests/webrtc/protocol/rtp-demuxing.html [ Crash ]
```
Phil had asked https://bugs.webkit.org/show_bug.cgi?id=286481 that when fixing the crash we restore the original expectation as it was failing, see the bug for details.

nikolaszimmermann replied:
Will do that later tonight.

csaavedra reviewed (commented):
LGTM, but I'm not a reviewer and I think this needs approval of a graphics dev.

aperezdc reviewed (approved):
The change makes sense to me. You might want to wait for another pair of eyes to take a look.

nikolaszimmermann committed:
[GTK] imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https.html is crashing in a Skia codepath
https://bugs.webkit.org/show_bug.cgi?id=304134

Reviewed by Adrian Perez de Castro.

When using Skia with GPU acceleration, NativeImages backed by textures must be
destroyed on the same thread where the Skia GrDirectContext was created, not on
the compositor thread. Releasing GPU resources on the wrong thread corrupts
Skia's GrResourceCache, leading to bounds check failures and crashes.

This patch ensures that in ~CoordinatedPlatformLayerBufferNativeImage(), if the
image is texture-backed, its destruction is dispatched to the main thread.

This also fixes crashes/assertions in:
- http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable.html
- http/wpt/mediastream/transfer-mediastreamtrack-to-worker.html
- imported/w3c/web-platform-tests/mst-content-hint/RTCRtpSendParameters-degradationEffect.html

Covered by existing tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedImageBackingStore.cpp:
(WebCore::CoordinatedImageBackingStore::isSameNativeImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.cpp:
(WebCore::CoordinatedPlatformLayerBufferNativeImage::~CoordinatedPlatformLayerBufferNativeImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.h:

Canonical link: https://commits.webkit.org/304897@main
